doc/vfontas.1
.TH vfontas 1 "2019-04-21" "hxtools" "hxtools"
.SH Name
vfontas \(em a bitmap font file transformation utility
.SH Introduction
vfontas (originally "VGA font file assembler") can read/write bitmap fonts
from/to a number of formats and transform the glyphs in various ways. vfontas
is able to generate outline fonts from bitmapped fonts.
.SH Syntax
\fBvfontas\fP \fIcommands\fP...
.SS Commands
\fB\-blankfnt\fP
.PP
\fB\-canvas\fP \fIxsize\fP \fIysize\fP
.PP
\fB\-clearmap\fP
.PP
\fB\-copy\fP \fIsrcx srcy width height dstx dsty\fP
.PP
\fB\-cpisep\fP \fIsep\fP
.PP
\fB\-crop\fP \fIxpos\fP \fIypos\fP \fIwidth\fP \fIheight\fP
.PP
\fB\-fliph\fP
.PP
\fB\-flipv\fP
.PP
\fB\-invert\fP
.PP
\fB\-lge\fP
.PP
\fB\-lgeu\fP
.PP
\fB\-lgeuf\fP
.PP
\fB\-loadbdf\fP \fIff.bdf\fP
.PP
\fB\-loadclt\fP \fIdirectory/\fP
.PP
\fB\-loadfnt\fP \fImu.fnt\fP
.PP
\fB\-loadhex\fP \fIunicode.hex\fP
.PP
\fB\-loadmap\fP \fIcp437AB.uni\fP
.PP
\fB\-loadpsf\fP \fIter-v32b.psfu\fP
.PP
\fB\-loadpsf <(gzip \-cd\fP \fIabc.psfu.gz\fP\fB)\fP
.PP
\fB\-loadraw\fP \fImu.fnt\fP \fIwidth\fP \fIheight\fP
.PP
\fB\-move\fP \fIshiftx\fP \fIshifty\fP
.PP
\fB\-overstrike\fP \fIxoffset\fP
.PP
\fB\-savebdf\fP \fIout.bdf\fP
.PP
\fB\-saveclt\fP \fIoutdir/\fP
.PP
\fB\-savefnt\fP \fIx.fnt\fP
.PP
\fB\-savemap\fP \fIchar2uni.map\fP
.PP
\fB\-saven1\fP \fInew.sfd\fP
.PP
\fB\-saven2\fP \fInew.sfd\fP
.PP
\fB\-savepbm\fP \fIoutdir/\fP
.PP
\fB\-savepsf\fP \fInew.psfu\fP
.PP
\fB\-savesfd\fP \fInew.sfd\fP
.PP
\fB\-setbold\fP
.PP
\fB\-setname\fP \fIname\fP
.PP
\fB\-setprop\fP \fIkey\fP \fIvalue\fP
.PP
\fB\-upscale\fP \fIxscale\fP \fIyscale\fP
.PP
\fB\-xcpi\fP \fIega.cpi\fP \fIoutdir/\fP
.PP
\fB\-xcpi.ice\fP \fIega.ice\fP \fIoutdir/\fP
.PP
\fB\-xlat\fP \fIxoffset\fP \fIyoffset\fP
.SH Formats
.IP \(bu 4
X11 BDF (Bitmap Distribution Font): read/write
.IP \(bu 4
vfontas CLT (Graphical Plaintext): read/write
.IP \(bu 4
Raw fonts used by e.g. DOS utilities: read/write
.IP \(bu 4
GNU .hex used by e.g. gnu-unifont: read-only
.IP \(bu 4
Linux/kbd PSF1 bitmaps: read-only
.IP \(bu 4
Linux/kbd PSF2 bitmaps: read/write
.IP \(bu 4
Fontforge SFD vectors: write-only
.IP \(bu 4
kbd Unicode maps: read/write
.IP \(bu 4
PSF2 Unicode maps: read/write
.IP \(bu 4
FreeBSD vfnt files: read-only
.IP \(bu 4
MS-DOS CPI container: read-only
.PP
vfontas does not have a direct integration of or with FontForge, but thanks to
the BDF read/write support on both sides, fonts in formats like FNT and PSF can
be made available to FontForge by vfontas, and, conversely, fonts in formats
like FON and TTF can be made available to vfontas by FontForge.
.PP
vfontas offers the CLT bitmap format, wherein glyphs are represented as
textgraphical files to facilitate editing with terminal editors.
.SH Multi-size bitmap fonts are styles
A bitmap font which is offered by its authors in multiple sizes can not only
constitute different sizes, but also visually different styles, similar to what
italic fonts are to regular fonts (e.g. longer descenders, more bulbous
terminals, etc.).
.PP
When vfontas converts bitmap fonts to vector, the "size" of a bitmap font
becomes "resolution". An 8x8 bitmap font and an 8x16 bitmap font, after
conversion to SFD and other vector formats \fBwill\fP have the same height when
the user requests such vectorized forms to be rendered in size 12pt.
.SH Vector font display sizes
Pixel-perfect rendering of a (vectorized) bitmap font requires taking the
original size in px and applying the monitor DPI. A 16px tall font on a system
which uses a 96dpi setting displays well using 12pt or multiples thereof.
(16*72/96=12). Here is a cheat table for some font heights:
.TS
allbox tab(:);
lB lB lB lB lB.
T{
BM Height
T}:T{
96dpi
T}:T{
x2
T}:T{
120dpi
T}:T{
x2
T}
.T&
l l l l.
T{
13px
T}:T{
9.75pt
T}:T{
19.5pt
T}:T{
7.8pt
T}:T{
15.6pt
T}
.T&
l l l l.
T{
14px
T}:T{
10.5pt
T}:T{
21pt
T}:T{
8.4pt
T}:T{
16.8pt
T}
.T&
l l l l.
T{
15px
T}:T{
11.25pt
T}:T{
22.5pt
T}:T{
9pt
T}:T{
18pt
T}
.T&
l l l l.
T{
16px
T}:T{
12pt
T}:T{
24pt
T}:T{
9.6pt
T}:T{
19.2pt
T}
.T&
l l l l.
T{
18px
T}:T{
13.5pt
T}:T{
27pt
T}:T{
10.8pt
T}:T{
21.6pt
T}
.T&
l l l l.
T{
20px
T}:T{
15pt
T}:T{
30pt
T}:T{
12pt
T}:T{
24pt
T}
.T&
l l l l.
T{
22px
T}:T{
16.5pt
T}:T{
T}:T{
13.2pt
T}:T{
26.4pt
T}
.T&
l l l l.
T{
24px
T}:T{
18pt
T}:T{
T}:T{
14.4pt
T}:T{
28.8pt
T}
.T&
l l l l.
T{
28px
T}:T{
21pt
T}:T{
T}:T{
16.8pt
T}:T{
T}
.T&
l l l l.
T{
32px
T}:T{
24pt
T}:T{
T}:T{
19.2pt
T}:T{
T}
.TE
.SH Aspect ratio
CRT screens of the time commonly had an aspect ratio of 4:3, and whatever
resolution was shown would be squeezed into that frame, potentially making the
pixels non-square. This means that bitmap fonts of the time will need a
corrective factor for modern display hardware to be reproduced faithfully. This
correction can happen either during conversion (vfontas has no support for this
at present), or at the time of rendering. The font can either be stretched in
the vertical direction or compressed in the horizontal by using the reciprocal.
.TS
allbox tab(:);
lB lB lB lB.
T{
Text res
T}:T{
Cell size
T}:T{
Resolution
T}:T{
VStretch
T}
.T&
l l l l.
T{
40x25
T}:T{
8x8
T}:T{
320x200
T}:T{
6/5 = 1.2
T}
.T&
l l l l.
T{
80x25
T}:T{
8x8
T}:T{
640x200
T}:T{
12/5 = 2.4
T}
.T&
l l l l.
T{
80x25
T}:T{
8x14
T}:T{
640x350
T}:T{
48/35 = 1.37
T}
.T&
l l l l.
T{
80x25
T}:T{
9x14
T}:T{
720x350
T}:T{
54/35 = 1.54
T}
.T&
l l l l.
T{
80x25
T}:T{
9x16
T}:T{
720x400
T}:T{
27/20 = 1.35
T}
.T&
l l l l.
T{
80x30
T}:T{
8x16
T}:T{
640x480
T}:T{
1.0
T}
.T&
l l l l.
T{
80x43
T}:T{
8x8
T}:T{
640x350
T}:T{
48/35 = 1.37
T}
.T&
l l l l.
T{
80x50
T}:T{
8x8
T}:T{
640x400
T}:T{
6/5 = 1.2
T}
.T&
l l l l.
::x*y:x/y*3/4
.TE
.SH Commands
.SS blankfnt
Initializes the memory buffer with 256 empty 8x16 glyphs. The primary purpose
for this is with \fBsaveclt\fP to get blank glyph files for hand-editing. For a
differently-sized canvas, combine with \fB\-crop\fP or \fB\-canvas\fP.
.SS canvas
Enlarges the glyph box to the specified size. (It never shrinks it.)
.SS clearmap
Discards the in-memory glyph index <-> Unicode mapping table.
.SS copy
Copy a portion of the bitmap from one place to another, overwriting pixels.
.SS cpisep
Switches to flat hierarchy extraction for \-xcpi*, using the specified character
for delimiter.
.SS crop
Removes an outer area from the glyph images, shrinking the image in the process.
.SS fliph, flipv
Mirrors/flips glyphs.
.SS lge
Applies a "Line Graphics Enable" transformation on glyphs. It copies the pixels
in the second rightmost column to the rightmost column, and does this for
\fIglyph indices\fP 0xC0 to 0xDF. In other words, the usefulness of the \-lge
command is more or less limited to DOS fonts which have graphic characters in
exactly those indices.
.SS lgeu
Applies LGE on the graphic glyphs that are \fIin cp437\fP and other DOS
codepages. It does this for \fIunicode codepoints\fP rather than glyph indices.
This is suitable for all kinds of font formats, but do make sure a Unicode
table is present. (BDF and PSF come with a table, .fnt does not, so use
\-loadmap before \-lgeu).
.SS lgeuf
Applies LGE on \fIall\fP glyphs from U+2500 thru U+25FF (Unicode "box drawing"
and "box elements" classes), with special handling for the shades at U+2591
thru U+2593. This is provided as an alternative to the "true DOS look" that
lge/lgeu would make.
.SS loadbdf
Reads a BDF (Adobe Glyph Bitmap Distribution Format) font file.
.SS loadclt
Reads a directory full of CLT files containing glyphs. CLT is a textgraphical
format to facilitate visual editing with a text console editor.
.SS loadfnt
Reads a headerless bitmap font file, as typically used for CGA/EGA/VGA/MDA
hardware, from the specified file into memory. 8x8x256 (width/height/glyphs),
8x12x256, 8x14x256, 8x16x256 and 8x16x512 are supported.

doc/NEWS.txt
rel-20201116
============
* pegrep: cure uninitialized variable warnings
* gpsh: change tmp location to /var/tmp
  (avoid saving potentially large files to tmpfs)
* sysinfo: fix null deref when /etc/os-release is absent


rel-20200712
============
* profile: add QUILT_COLORS
* bsvplay: Pianoman MUS support
* pcmmix: new utility for arpeggio/polyphonic mixing of
  bsvplay/qplay outputs


rel-20200428
============
* vfontas: N2 vectorizer: cease smoothing of stub serifs
  (concerns the topleft corner of an 'E' in some fonts)
* vfontas: N2 vectorizer: avoid diagonalizing "pimples" in font outlines
  (concerns 1-pixel protrusions in the right vertical edge of the '4')


rel-20200331
============
* vfontas: add new N2 vectorizer
* vfontas: new -saven2 command to use N2

The N2 vectorizer is able to recognize logical 45-degree angles in
bitmap glyphs and will output the outlines as such. Unlike N1,
it analyzes glyph outlines rather than pixel patterns, and produces
higher quality output due to a larger working radius.

N2 can reproduce the "Nouveau IBM" font quite closely now (e.g. from
Fontraption's VGA9.F16 file). A few differences have just become apparent (one
can inspect in e.g. Inkscape by overlaying both fontfaces) that may be
addressed in a later release.

N2:
* Isthmuses tend to get washed out, noticable in glyphs like 'G'.
* Sumps in the 'K', 'M', 'N' glyphs could be improved.
  This is noticable in glyphs like 'G'.

Nouveau IBM:
* Kerning is not original: too closely spaced.
* Node points are not at pixel or half-pixel locations.
  'B' has a 30.78 diagonal for no apparent reason. This is bad for integral
  scaling. Other values can be observed in 'W'.
* Glyph 'C' is asymmetric.


rel-20200309
============
* vfontas: add N1 vectorizer
* vfontas: new -saven1 command to use N1
* vfontas: remove -xbrz



rel-20200124
============
* vfontas: add full vectorizer for SFD glyphs
* vfontas: improved ascent/descent values for vector output
* vfontas: hex format misreading was fixed


rel-20190428
============
* vfontas: new implementation. Replaces clt2bdf, clt2pbm, cltxscale, fnt2bdf.
* vfontas: xBRZ x6 support
* profile: use new GREP_COLORS variable


rel-20180301
====


src/vfontas.cpp
// SPDX-License-Identifier: GPL-3.0-or-later
// SPDX-FileCopyrightText: 2019–2025 Jan Engelhardt
/*
 *	Command-line interface of the "VGA font assembler"
 */
#include "config.h"
#include <algorithm>
#include <memory>
#include <string>
#include <utility>
#include <cstdint>
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <fcntl.h>
#include <unistd.h>
#include <sys/mman.h>
#include <sys/stat.h>
#include <libHX/defs.h>
#include <libHX/io.h>
#include <libHX/string.h>
#include "vfalib.hpp"
#define MMAP_NONE reinterpret_cast<void *>(-1)

using namespace vfalib;

/* CPI: see http://www.seasip.info/DOS/CPI/cpi.html */
struct cpi_fontfile_header {
	uint8_t id0;
	char id[7], reserved[8];
	uint16_t pnum;
	uint8_t ptyp;
	uint32_t fih_offset;
} __attribute__((packed));

struct cpi_fontinfo_header {
	uint16_t num_codepages;
} __attribute__((packed));

enum cpi_device_type {
	DEVTYPE_SCREEN = 1,
	DEVTYPE_PRINTER,
};

/**
 * @device_name:	for screens, usually "EGA", or perhaps "LCD".
 * 			for printers, usually "4201", "4208", "5202",
 * 			"1050", "EPS", "PPDS"
 */
struct cpi_cpentry_header {
	uint16_t cpeh_size;
	uint32_t next_cpeh_offset;
	uint16_t device_type;
	char device_name[8];
	uint16_t codepage;
	char reserved[6];
	uint32_t cpih_offset;
} __attribute__((packed));

struct cpi_cpinfo_header {
	uint16_t version;
	uint16_t num_fonts;
	uint16_t size;
} __attribute__((packed));

struct cpi_screenfont_header {
	uint8_t height, width, yaspect, xaspect;
	uint16_t num_chars;
} __attribute__((packed));

struct cpi_printfont_header {
	uint16_t printer_type, escape_length;
} __attribute__((packed));

static std::string cpi_separator;

static bool vf_blankfnt(font &f, char **args)
{
	f.init_256_blanks();
	return true;
}

static bool vf_canvas(font &f, char **args)
{
	auto x = strtol(args[0], nullptr, 0);
	auto y = strtol(args[1], nullptr, 0);
	if (x < 0 || y < 0) {
		fprintf(stderr, "Error: Canvas size should be positive.\n");
		return false;
	}
	if (f.m_glyph.size() > 0)
		f.copy_to_blank(vfpos() | f.m_glyph[0].m_size, vfpos() | vfsize(x, y));
	return true;
}

static bool vf_clearmap(font &f, char **args)
{
	f.m_unicode_map.reset();
	return true;
}

static bool vf_copy(font &f, char **args)
{
	auto x = strtol(args[0], nullptr, 0);
	auto y = strtol(args[1], nullptr, 0);
	auto w = strtol(args[2], nullptr, 0);
	auto h = strtol(args[3], nullptr, 0);
	auto bx = strtol(args[4], nullptr, 0);
	auto by = strtol(args[5], nullptr, 0);
	if (x < 0 || y < 0) {
		fprintf(stderr, "Error: Crop xpos/ypos must be positive.\n");
		return false;
	}
	if (w <= 0 || h <= 0) {
		fprintf(stderr, "Error: Crop width/height must be positive non-zero.\n");
		return false;
	}
	if (f.m_glyph.size() > 0)
		f.copy_rect(vfpos(x, y) | vfsize(w, h), vfpos(bx, by) | f.m_glyph[0].m_size);
	return true;
}

static bool vf_cpisep(font &f, char **args)
{
	cpi_separator = args[0];
	return true;
}

static bool vf_crop(font &f, char **args)
{
	auto x = strtol(args[0], nullptr, 0);
	auto y = strtol(args[1], nullptr, 0);
	auto w = strtol(args[2], nullptr, 0);
	auto h = strtol(args[3], nullptr, 0);
	if (x < 0 || y < 0) {
		fprintf(stderr, "Error: Crop xpos/ypos must be positive.\n");
		return false;
	}
	if (w <= 0 || h <= 0) {
		fprintf(stderr, "Error: Crop width/height must be positive non-zero.\n");
		return false;
	}
	if (f.m_glyph.size() > 0)
		f.copy_to_blank(vfpos(x, y) | f.m_glyph[0].m_size, vfpos() | vfsize(w, h));
	return true;
}

static bool vf_fliph(font &f, char **args)
{
	f.flip(true, false);
	return true;
}

static bool vf_flipv(font &f, char **args)
{
	f.flip(false, true);
	return true;
}

static bool vf_invert(font &f, char **args)
{
	f.invert();
	return true;
}

static bool vf_lge(font &f, char **args)
{
	f.lge();
	return true;
}

static bool vf_lgeu(font &f, char **args)
{
	f.lgeu();
	return true;
}

static bool vf_lgeuf(font &f, char **args)
{
	f.lgeuf();
	return true;
}

static bool vf_loadbdf(font &f, char **args)
{
	auto ret = f.load_bdf(args[0]);
	if (ret >= 0)
		return true;
	fprintf(stderr, "Error loading %s: %s\n", args[0], strerror(-ret));
	return false;
}

static bool vf_loadclt(font &f, char **args)
{
	auto ret = f.load_clt(args[0]);
	if (ret >= 0)
		return true;
	fprintf(stderr, "Error loading %s: %s\n", args[0], strerror(-ret));
	return false;
}

static bool vf_loadfnt(font &f, char **args)
{
	auto ret = f.load_fnt(args[0]);
	if (ret >= 0)
		return true;
	fprintf(stderr, "Error loading %s: %s\n", args[0], strerror(-ret));
	return false;
}

static bool vf_loadraw(font &f, char **args)
{
	auto ret = f.load_fnt(args[0], atoi(args[1]), atoi(args[2]));
	if (ret >= 0)
		return true;
	fprintf(stderr, "Error loading %s: %s\n", args[0], strerror(-ret));
	return false;
}

static bool vf_loadhex(font &f, char **args)
{
	auto ret = f.load_hex(args[0]);
	if (ret >= 0)
		return true;
	fprintf(stderr, "Error loading %s: %s\n", args[0], strerror(-ret));
	return false;
}

static bool vf_loadmap(font &f, char **args)
{
	if (f.m_unicode_map == nullptr)
		f.m_unicode_map = std::make_shared<unicode_map>();
	auto ret = f.m_unicode_map->load(args[0]);
	if (ret >= 0)
		return true;
	fprintf(stderr, "Error loading %s: %s\n", args[0], strerror(-ret));
	return false;
}

static bool vf_loadpcf(font &f, char **args)
{
	auto ret = f.load_pcf(args[0]);
	if (ret >= 0)
		return true;
	fprintf(stderr, "Error loading %s: %s\n", args[0], strerror(-ret));
	return false;
}

static bool vf_loadpsf(font &f, char **args)
{
	auto ret = f.load_psf(args[0]);
	if (ret >= 0)
		return true;
	fprintf(stderr, "Error loading %s: %s\n", args[0], strerror(-ret));
	return false;
}

static bool vf_loadvfnt(font &f, char **args)
{
	auto ret = f.load_vfnt(args[0]);
	if (ret >= 0)
		return true;
	fprintf(stderr, "Error loading %s: %s\n", args[0], strerror(-ret));
	return false;
}

static bool vf_move(font &f, char **args)
{
	auto x = strtol(args[0], nullptr, 0);
	auto y = strtol(args[1], nullptr, 0);
	if (f.m_glyph.size() <= 0)
		return true;
	f.copy_to_blank(vfpos() | f.m_glyph[0].m_size, vfpos(x, y) | f.m_glyph[0].m_size);
	return true;
}

static bool vf_overstrike(font &f, char **args)
{
	f.overstrike(strtoul(args[0], nullptr, 0));
	return true;
}

static bool vf_savebdf(font &f, char **args)
{
	auto ret = f.save_bdf(args[0]);
	if (ret >= 0)
		return true;
	fprintf(stderr, "Error saving %s: %s\n", args[0], strerror(-ret));
	return false;
}

static bool vf_saveclt(font &f, char **args)
{
	auto ret = f.save_clt(args[0]);
	if (ret >= 0)
		return true;
	fprintf(stderr, "Error saving %s/: %s\n", args[0], strerror(-ret));
	return false;
}

static bool vf_savefnt(font &f, char **args)
{
	auto ret = f.save_fnt(args[0]);
	if (ret >= 0)
		return true;
	fprintf(stderr, "Error saving %s: %s\n", args[0], strerror(-ret));
	return false;
}

static bool vf_savemap(font &f, char **args)
{
	auto ret = f.save_map(args[0]);
	if (ret >= 0)
		return true;
	fprintf(stderr, "Error saving %s: %s\n", args[0], strerror(-ret));
	return false;
}

static bool vf_savepbm(font &f, char **args)
{
	auto ret = f.save_pbm(args[0]);
	if (ret >= 0)
		return true;
	fprintf(stderr, "Error saving %s: %s\n", args[0], strerror(-ret));
	return false;
}

static bool vf_savepsf(font &f, char **args)
{
	auto ret = f.save_psf(args[0]);
	if (ret >= 0)
		return true;
	fprintf(stderr, "Error saving %s: %s\n", args[0], strerror(-ret));
	return false;
}

static bool vf_savesfd(font &f, char **args)
{
	auto ret = f.save_sfd(args[0], vectoalg::V_SIMPLE);
	if (ret >= 0)
		return true;
	fprintf(stderr, "Error saving %s: %s\n", args[0], strerror(-ret));
	return false;
}

static bool vf_saven1(font &f, char **args)
{
	auto ret = f.save_sfd(args[0], vectoalg::V_N1);
	if (ret >= 0)
		return true;
	fprintf(stderr, "Error saving %s: %s\n", args[0], strerror(-ret));
	return false;
}

static bool vf_saven2(font &f, char **args)
{
	auto ret = f.save_sfd(args[0], vectoalg::V_N2);
	if (ret >= 0)
		return true;
	fprintf(stderr, "Error saving %s: %s\n", args[0], strerror(-ret));
	return false;
}

static bool vf_saven2ev(font &f, char **args)
{
	auto ret = f.save_sfd(args[0], vectoalg::V_N2EV);
	if (ret >= 0)
		return true;
	fprintf(stderr, "Error saving %s: %s\n", args[0], strerror(-ret));
	return false;
}

static bool vf_setbold(font &f, char **args)
{
	f.props.insert_or_assign("TTFWeight", "700");
	f.props.insert_or_assign("StyleMap", "0x0020");
	f.props.insert_or_assign("Weight", "bold");
	return true;
}

static bool vf_setname(font &f, char **args)
{
	std::string ps_name = args[0];
	/* PostScript name does not allow spaces */
	std::replace(ps_name.begin(), ps_name.end(), ' ', '-');
	f.props.insert_or_assign("FontName", std::move(ps_name));
	f.props.insert_or_assign("FullName", args[0]);
	f.props.insert_or_assign("FamilyName", args[0]);
	f.props.emplace("Weight", "medium");
	return true;
}

static bool vf_setprop(font &f, char **args)
{
	f.props.insert_or_assign(args[0], args[1]);
	return true;
}

static bool vf_upscale(font &f, char **args)
{
	auto xf = strtol(args[0], nullptr, 0);
	auto yf = strtol(args[1], nullptr, 0);
	if (xf <= 0 || yf <= 0) {
		fprintf(stderr, "Error: scaling factor(s) should be positive and greater than zero.\n");
		return false;
	}
	f.upscale(vfsize(xf, yf));
	return true;
}

static uint32_t xlate_segoff(uint32_t x)
{
	return (x >> 12) + (x & 0xFFFF);
}

static void vf_extract_sfh(const char *sfhblk, unsigned int num_fonts,
    const std::string &tpl_dir, const char *dev, const char *cpg)
{
	for (unsigned int i = 0; i < num_fonts; ++i) {
		struct cpi_screenfont_header sfh;
		memcpy(&sfh, sfhblk, sizeof(sfh));
		sfh.num_chars = std::min(le16_to_cpu(sfh.num_chars), static_cast<uint16_t>(UINT16_MAX));
		printf("SFH: %ux%u pixels x %u chars\n", sfh.width, sfh.height,
		       sfh.num_chars);
		if (sfh.width == 0 || sfh.height == 0 || sfh.num_chars == 0)
			/* Avoid producing empty files */
			continue;

		char buf[HXSIZEOF_Z32*3];
		snprintf(buf, sizeof(buf), "%ux%u.fnt", sfh.width, sfh.height);
		auto out_dir = tpl_dir;
		auto out_file = tpl_dir;
		if (cpi_separator.size() > 0) {
			out_file += std::string("/") + dev + cpi_separator +
			            cpg + cpi_separator + buf;
		} else {
			out_dir += std::string("/") + dev + "/" + cpg;
			out_file = out_dir + "/" + buf;
		}
		printf("Writing to %s\n", out_file.c_str());
		HX_mkdir(out_dir.c_str(), S_IRWXUGO);
		auto out_fd = open(out_file.c_str(), O_WRONLY | O_CREAT | O_TRUNC, S_IRUGO | S_IWUGO);
		if (out_fd < 0) {
			fprintf(stderr, "Error writing to %s: %s\n", out_file.c_str(), strerror(errno));
			continue;
		}

		auto length = sfh.width * sfh.height / 8 * sfh.num_chars;
		write(out_fd, sfhblk + sizeof(sfh), length);
		close(out_fd);
		sfhblk += sizeof(sfh) + length;
	}
}

static void vf_extract_pfh(const char *pfhblk)
{
	struct cpi_printfont_header pfh;
	memcpy(&pfh, pfhblk, sizeof(pfh));
	pfh.printer_type = le16_to_cpu(pfh.printer_type);
	pfh.escape_length = le16_to_cpu(pfh.escape_length);
	printf("PFH: printer_type=%u escape_len=%u\n",
	       pfh.printer_type, pfh.escape_length);
}

static int vf_extract_cpi2(const char *vdata, size_t vsize,
    const char *directory, bool seg_mode)
{
	struct cpi_fontfile_header ffh;
	struct cpi_fontinfo_header fih;

	memcpy(&ffh, vdata, sizeof(ffh));
	ffh.pnum = le16_to_cpu(ffh.pnum);
	ffh.fih_offset = le32_to_cpu(ffh.fih_offset);

	if (ffh.id0 != 0xFF || strncmp(ffh.id, "FONT    ", sizeof(ffh.id)) != 0 ||
	    ffh.pnum != 1 || ffh.ptyp != 1)
		return -EINVAL;
	if (ffh.fih_offset + sizeof(fih) >= vsize)
		return -EINVAL;

	memcpy(&fih, vdata + ffh.fih_offset, sizeof(fih));
	fih.num_codepages = std::min(le16_to_cpu(fih.num_codepages), static_cast<uint16_t>(UINT16_MAX));

	const char *cpeblk = vdata + ffh.fih_offset + sizeof(fih);
	for (unsigned int i = 0; i < fih.num_codepages; ++i) {
		struct cpi_cpentry_header cpeh;
		struct cpi_cpinfo_header cpih;

		if (cpeblk - vdata + sizeof(cpeh) >= vsize)
			return -EINVAL;
		memcpy(&cpeh, cpeblk, sizeof(cpeh));
		cpeh.cpeh_size        = le16_to_cpu(cpeh.cpeh_size);
		if (cpeh.cpeh_size != sizeof(cpeh))
			return -EINVAL;
		cpeh.next_cpeh_offset = seg_mode ?
		                        xlate_segoff(le32_to_cpu(cpeh.next_cpeh_offset)) :
		                        le32_to_cpu(cpeh.next_cpeh_offset);
		cpeh.device_type      = le16_to_cpu(cpeh.device_type);
		cpeh.codepage         = le16_to_cpu(cpeh.codepage);
		cpeh.cpih_offset      = seg_mode ?
		                        xlate_segoff(le32_to_cpu(cpeh.cpih_offset)) :
		                        le32_to_cpu(cpeh.cpih_offset);
		cpeblk = vdata + cpeh.next_cpeh_offset;

		printf("CPEH #%u: Name: %.*s, Codepage: %u, Device: %.*s, DType: %u\n",
		       i, static_cast<int>(sizeof(cpeh.device_name)),
		       cpeh.device_name, cpeh.codepage,
		       static_cast<int>(std::size(cpeh.device_name)),
		       cpeh.device_name, cpeh.device_type);

		if (cpeh.next_cpeh_offset + sizeof(cpeh) >= vsize)
			return -EINVAL;
		if (cpeh.cpih_offset + sizeof(cpih) >= vsize)
			return -EINVAL;
		memcpy(&cpih, vdata + cpeh.cpih_offset, sizeof(cpih));
		cpih.version   = le16_to_cpu(cpih.version);
		cpih.num_fonts = std::min(le16_to_cpu(cpih.num_fonts), static_cast<uint16_t>(UINT16_MAX));
		cpih.size      = le16_to_cpu(cpih.size);
		printf("CPIH: version=%u fonts=%u size=%u\n", cpih.version,
		       cpih.num_fonts, cpih.size);
		if (cpih.version != 1)
			continue;

		char dev[HXSIZEOF_Z32*2], cpg[HXSIZEOF_Z32*2];
		*dev = '\0';
		HX_strlncat(dev, cpeh.device_name, std::size(dev), std::size(cpeh.device_name));
		HX_strrtrim(dev);
		snprintf(cpg, std::size(cpg), "%u", cpeh.codepage);
		if (cpeh.device_type == DEVTYPE_SCREEN)
			vf_extract_sfh(vdata + cpeh.cpih_offset + sizeof(cpih),
				cpih.num_fonts, directory, dev, cpg);
		else if (cpeh.device_type == DEVTYPE_PRINTER)
			vf_extract_pfh(vdata + cpeh.cpih_offset + sizeof(cpih));
	}
	return true;
}

static bool vf_xcpi(font &f, char **args, bool seg_mode)
{
	auto in_fd = open(args[0], O_RDONLY);
	if (in_fd < 0) {
		fprintf(stderr, "Could not open %s: %s\n", args[0], strerror(errno));
		return false;
	}
	auto fdclean = make_scope_success([&]() { close(in_fd); });
	struct stat sb;
	if (fstat(in_fd, &sb) < 0) {
		fprintf(stderr, "fstat: %s\n", strerror(errno));
		return false;
	}

	auto mapping = mmap(nullptr, sb.st_size, PROT_READ, MAP_SHARED, in_fd, 0);
	if (mapping == MMAP_NONE) {
		fprintf(stderr, "mmap: %s\n", strerror(errno));
		return false;
	}
	auto mapclean = make_scope_success([&]() { munmap(mapping, sb.st_size); });
	auto ret = vf_extract_cpi2(static_cast<const char *>(mapping),
	           sb.st_size, args[1], seg_mode);
	if (ret == -EINVAL) {
		fprintf(stderr, "xcpi: file \"%s\" not recognized\n", args[0]);
		return false;
	}
	return ret >= 0;
}

static bool vf_xcpi_flat(font &f, char **args)
{
	return vf_xcpi(f, args, false);
}

static bool vf_xcpi_seg(font &f, char **args)
{
	return vf_xcpi(f, args, true);
}

static bool vf_xlat(font &f, char **args)
{
	auto x = strtol(args[0], nullptr, 0);
	auto y = strtol(args[1], nullptr, 0);
	if (f.m_glyph.size() > 0)
		f.copy_to_blank(vfpos() | f.m_glyph[0].m_size, vfpos(x, y) | f.m_glyph[0].m_size);
	return true;
}

static const struct vf_command {
	const char *cmd;
	unsigned int nargs;
	bool (*func)(font &f, char **args);
} vf_commlist[] = {
	{"blankfnt", 0, vf_blankfnt},
	{"canvas", 2, vf_canvas},
	{"clearmap", 0, vf_clearmap},
	{"copy", 6, vf_copy},
	{"cpisep", 1, vf_cpisep},
	{"crop", 4, vf_crop},
	{"fliph", 0, vf_fliph},
	{"flipv", 0, vf_flipv},
	{"invert", 0, vf_invert},
	{"lge", 0, vf_lge},
	{"lgeu", 0, vf_lgeu},
	{"lgeuf", 0, vf_lgeuf},
	{"loadbdf", 1, vf_loadbdf},
	{"loadclt", 1, vf_loadclt},
	{"loadfnt", 1, vf_loadfnt},
	{"loadhex", 1, vf_loadhex},
	{"loadmap", 1, vf_loadmap},
	{"loadpcf", 1, vf_loadpcf},
	{"loadpsf", 1, vf_loadpsf},
	{"loadraw", 3, vf_loadraw},
	{"loadvfnt", 1, vf_loadvfnt},
	{"move", 2, vf_move},
	{"overstrike", 1, vf_overstrike},
	{"savebdf", 1, vf_savebdf},
	{"saveclt", 1, vf_saveclt},
	{"savefnt", 1, vf_savefnt},
	{"savemap", 1, vf_savemap},
	{"saven1", 1, vf_saven1},
	{"saven2", 1, vf_saven2},
	{"saven2ev", 1, vf_saven2ev},
	{"savepbm", 1, vf_savepbm},
	{"savepsf", 1, vf_savepsf},
	{"savesfd", 1, vf_savesfd},
	{"setbold", 0, vf_setbold},
	{"setname", 1, vf_setname},
	{"setprop", 2, vf_setprop},
	{"upscale", 2, vf_upscale},
	{"xcpi", 2, vf_xcpi_flat},
	{"xcpi.ice", 2, vf_xcpi_seg},
	{"xlat", 2, vf_xlat},
};

int main(int argc, char **argv)
{
	font f;
	if (argc > 0) {
		--argc;
		++argv;
	}
	while (argc > 0) {
		if (argv[0][0] == '-')
			++argv[0];
		auto ce = static_cast<const vf_command *>(bsearch(argv[0], vf_commlist, ARRAY_SIZE(vf_commlist), sizeof(*vf_commlist),
			[](const void *cmd, const void *arr) -> int {
				return strcmp(static_cast<const char *>(cmd), static_cast<const vf_command *>(arr)->cmd);
			}));
		if (ce == nullptr) {
			fprintf(stderr, "Error: Unknown command \"%s\"\n", argv[0]);
			return EXIT_FAILURE;
		}
		--argc;
		if (static_cast<unsigned int>(argc) < ce->nargs) {
			fprintf(stderr, "Error: Command \"%s\" requires %u arguments.\n", argv[0], ce->nargs);
			return EXIT_FAILURE;
		}
		if (!ce->func(f, ++argv))
			return EXIT_FAILURE;
		argc -= ce->nargs;
		argv += ce->nargs;
	}
	return EXIT_SUCCESS;
}
